sdk/core/core/src/az_json_read.c

```diff
@@ -0,0 +1,379 @@
+              return AZ_JSON_ERROR_UNEXPECTED_CHAR;
+            }
+          }
+        }
+    }
+    if (c < 0x20) {
+      return AZ_JSON_ERROR_UNEXPECTED_CHAR;
+    }
+    *p += 1;
+  }
+}
+
+// _value_
+az_result az_json_read_value(az_json_state *const p_state, az_json_value *const out_value) {
+  az_const_str const buffer = p_state->buffer;
+  size_t *const p = &p_state->i;
+  if (*p == buffer.size) {
+	  return AZ_JSON_ERROR_UNEXPECTED_END;
+  }
+  char const c = az_const_str_item(buffer, *p);
+  if (az_is_digit(c)) {
+    out_value->tag = AZ_JSON_NUMBER;
+    return az_json_read_number_digit_rest(buffer, p, &out_value->number, 1);
+  }
+  switch (c) {
+    case 't':
+      out_value->tag = AZ_JSON_BOOLEAN;
+      out_value->boolean = true;
+      return az_json_read_keyword_rest(buffer, p, AZ_CONST_STR("rue"));
+    case 'f':
+      out_value->tag = AZ_JSON_BOOLEAN;
+      out_value->boolean = false;
+      return az_json_read_keyword_rest(buffer, p, AZ_CONST_STR("alse"));
+    case 'n':
+      out_value->tag = AZ_JSON_NULL;
+      return az_json_read_keyword_rest(buffer, p, AZ_CONST_STR("ull"));
+    case '"':
+      out_value->tag = AZ_JSON_STRING;
+      *p += 1;
+      return az_json_read_string_rest(buffer, p, &out_value->string);
+    case '-':
+      out_value->tag = AZ_JSON_NUMBER;
+      return az_json_read_number_minus_rest(buffer, p, &out_value->number);
+    case '{':
+      out_value->tag = AZ_JSON_OBJECT;
+      *p += 1;
+      return az_json_stack_push(p_state, AZ_JSON_STACK_OBJECT);
+    case '[':
+      out_value->tag = AZ_JSON_ARRAY;
+      *p += 1;
+      return az_json_stack_push(p_state, AZ_JSON_STACK_ARRAY);
+  }
+  return AZ_JSON_ERROR_UNEXPECTED_CHAR;
+}
+
+az_result az_json_read_value_space(az_json_state *const p_state, az_json_value *const out_value) {
+  AZ_RETURN_ON_ERROR(az_json_read_value(p_state, out_value));
+  return az_json_read_white_space(p_state->buffer, &p_state->i);
+}
+
+az_result az_json_read(az_json_state *const p_state, az_json_value *const out_value) {
+  if (!az_json_stack_is_empty(p_state)) {
+    return AZ_JSON_ERROR_INVALID_STATE;
+  }
+  AZ_RETURN_ON_ERROR(az_json_read_white_space(p_state->buffer, &p_state->i));
+  AZ_RETURN_ON_ERROR(az_json_read_value_space(p_state, out_value));
+  bool const is_empty = p_state->i == p_state->buffer.size;
+  switch (out_value->tag) {
+    case AZ_JSON_ARRAY:
+    case AZ_JSON_OBJECT:
+      return is_empty ? AZ_JSON_ERROR_UNEXPECTED_END : AZ_OK;
+  }
+  return is_empty ? AZ_OK: AZ_JSON_ERROR_UNEXPECTED_CHAR;
+}
+
+az_result az_json_read_comma(az_json_state* const p_state) {
+  char c;
+  AZ_RETURN_ON_ERROR(az_json_get_char(p_state->buffer, p_state->i, &c));
+  if (c == ',') {
+    // skip ',' and read all whitespaces.
+    ++p_state->i;
+    return az_json_read_white_space(p_state->buffer, &p_state->i);
+  }
+  char const close = az_json_stack_last(p_state) == AZ_JSON_STACK_OBJECT ? '}' : ']';
```

Comment:
please use an if statement. I don't like ?:s with more than one "thing" in each part.